Crashlytics/Crashlytics/Rollouts/CrashlyticsRemoteConfigManager.swift

```diff
@@ -15,38 +15,56 @@
 import FirebaseRemoteConfigInterop
 import Foundation
 
-protocol CrashlyticsPersistentLog: NSObject {
-  func updateRolloutsStateToPersistence(rolloutAssignments: [RolloutAssignment])
+@objc(FIRCLSPersistenceLog)
+public protocol CrashlyticsPersistenceLog {
+  func updateRolloutsStateToPersistence(rollouts: Data, reportID: String)
 }
 
 @objc(FIRCLSRemoteConfigManager)
 public class CrashlyticsRemoteConfigManager: NSObject {
   public static let maxRolloutAssignments = 128
   public static let maxParameterValueLength = 256
 
+  private let lock = NSLock()
+  private var _rolloutAssignment: [RolloutAssignment] = []
+
   var remoteConfig: RemoteConfigInterop
-  @objc public private(set) var rolloutAssignment: [RolloutAssignment] = []
-  weak var persistenceDelegate: CrashlyticsPersistentLog?
+  var persistenceDelegate: CrashlyticsPersistenceLog
 
-  @objc public init(remoteConfig: RemoteConfigInterop) {
-    self.remoteConfig = remoteConfig
+  @objc public var rolloutAssignment: [RolloutAssignment] {
+    lock.lock()
+    defer { lock.unlock() }
+    let copy = _rolloutAssignment
+    return copy
   }
 
-  @objc public func updateRolloutsState(rolloutsState: RolloutsState) {
-    rolloutAssignment = normalizeRolloutAssignment(assignments: Array(rolloutsState.assignments))
+  @objc public init(remoteConfig: RemoteConfigInterop,
+                    persistenceDelegate: CrashlyticsPersistenceLog) {
+    self.remoteConfig = remoteConfig
+    self.persistenceDelegate = persistenceDelegate
   }
 
-  @objc public func getRolloutAssignmentsEncodedJson() -> String? {
-    let contentEncodedRolloutAssignments = rolloutAssignment.map { assignment in
-      EncodedRolloutAssignment(assignment: assignment)
+  @objc public func updateRolloutsState(rolloutsState: RolloutsState, reportID: String) {
+    lock.lock()
+    _rolloutAssignment = normalizeRolloutAssignment(assignments: Array(rolloutsState.assignments))
+    lock.unlock()
+
+    // writring to persistence
+    if let rolloutsData =
```

Comment:
Maybe a nit, but let's say at one moment there are rollouts, and then they're all deleted so there are no rollouts. Would that prevent us from notifying the delegate? Should we maybe always return a string and just make it empty if it fails to encode?

Reply:
If there is no rollouts I think encoder will encode as `{"rollouts": []}` and pass that to persistence. I tend to leverage on the optional for encode check to prevent we pass corrupted json to record.

Reply:
Makes sense 👍